src/NSrtm.Core/BiCubicInterpolation/Spline3DCalculator.cs

```diff
@@ -0,0 +1,145 @@
+        {
+            return (next - previous) / (2 * step);
+        }
+
+        private static List<double> getCoefficients(List<double> x)
+        {
+            var coefficients = new List<double>();
+
+            for (int i = 0; i < _linearEquationCoefficients.GetLength(0); i++)
+            {
+                double coefficient = x.Select((t, j) => _linearEquationCoefficients[i, j] * t)
+                                      .Sum();
+                coefficients.Add(coefficient);
+            }
+            return coefficients;
+        }
+
+        private static List<double> getParametersDescribingGrid(IReadOnlyList<IReadOnlyList<double>> values, double step)
+        {
+            var firstDerivativeX = values.Select(row => numericalDifferentation(row, step))
+                                         .ToList()
+                                         .AsReadOnly();
+
+            var firstDerivativeY = values.Select((t, i) => values.Select(element => element[i]))
+                                         .Select(column => numericalDifferentation(column.ToList(), step))
+                                         .ToList()
+                                         .AsReadOnly();
+            var crossDerivative = new List<List<double>>();
+            for (int i = 0; i < firstDerivativeY[0].Count; i++)
+            {
+                var col = firstDerivativeY.Select(p => p[i]);
+                crossDerivative.Add(numericalDifferentation(col.ToList(), step));
+            }
+
+            var x = new List<double>
+                    {
+                        values[1][1],
+                        values[1][2],
+                        values[2][1],
+                        values[2][2],
+                        firstDerivativeX[1][0],
+                        firstDerivativeX[1][1],
+                        firstDerivativeX[2][0],
+                        firstDerivativeX[2][1],
+                        firstDerivativeY[1][0],
+                        firstDerivativeY[1][1],
+                        firstDerivativeY[2][0],
+                        firstDerivativeY[2][1],
+                        crossDerivative[0][0],
+                        crossDerivative[0][1],
+                        crossDerivative[1][0],
+                        crossDerivative[1][1]
+                    };
+            return x;
+        }
+
+        private static double getFromPoly(List<double> coefficients, double xPos, double yPos)
+        {
+            double result = 0;
+            for (int i = 0; i < 4; i++)
+            {
+                for (int j = 0; j < 4; j++)
+                {
+                    result += coefficients[i * 4 + j] * Math.Pow(xPos, j) * Math.Pow(yPos, i);
+                }
+            }
+            return result;
+        }
+
+        #region Static Members
+
+        private static readonly int[,] _linearEquationCoefficients =
```

Comment:
Using jagged array would be a bit faster, but please supress warning with valid comment.